### PR TITLE
fix JENKINS-62407: Builds displaying "downstreamPipelineTriggerRunListener" when not configured to do so

### PR DESCRIPTION
**Pull request by bguerin**
Fix https://issues.jenkins.io/browse/JENKINS-62407

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by bguerin** [comment]
@aheritier Can I have your thought about this please. I know that my `MavenPipelineMarkerAction` is a hack, but there is no API to find executed steps on a `WorkflowRun`, so I did not know how to do different

**Comment by aheritier** [comment]
> there is no API to find executed steps on a WorkflowRun, so I did not know how to do different

Effectively I am not sure that there is a cleaner solution today. @jglick @Dohbedoh do you have a better idea ?

**Comment by Dohbedoh** [comment]
While this looks like a hack, this could well help mitigate the overhead of [scanning through the FlowGraph](https://github.com/jenkinsci/workflow-api-plugin/blob/7f5f1759dc1635f4e704d26e7bcba37a20e59c4b/docs/flowgraph.md#working-with-the-flowgraph). Would need a core / pipeline maintainer to confirm though. 

cc @jenkinsci/workflow-api-plugin-developers

**Comment by bguerin** [comment]
> > there is no API to find executed steps on a `WorkflowRun`
> 
> Sure there is. Start by checking for (`build instanceof`) https://javadoc.jenkins.io/plugin/workflow-api/org/jenkinsci/plugins/workflow/flow/FlowExecutionOwner.Executable.html (or if you already have a `WorkflowRun` via dep on `workflow-job`, which is a bad practice, call those methods directly) and you can iterate the flow graph looking for steps.

Many thanks @jglick 
I will go that way

Sorry for my statement, no offense, I said that because
https://www.jenkins.io/doc/developer/plugin-development/pipeline-integration/#run-listeners-vs-publishers
> [...] in the case of Pipeline there is no way to introspect the “list of build steps”

I found that weird but as it is in the official dev docs

**Comment by jglick** [comment]
There _is_ no way to introspect the list of build steps for a _job_. This could change from build to build, in completely unpredictable ways, at least in Scripted syntax (or even “Declarative” using libraries or `script` blocks). A perverse `Jenkinsfile` could read

```groovy
if (shouldDeployToProduction(BUILD_NUMBER as int)) {
  node {
    withMaven {
      sh 'mvn deploy'
    }
  }
}
@NonCPS boolean shouldDeployToProduction(int buildNumber) {
  int n = buildNumber * 2 + 2
  for (int i = 2; i < n; i++) {
    if (isPrime(i) && isPrime(n - i)) {
      return true
    }
  }
  return false
}
@NonCPS boolean isPrime(int n) {/* … */}
```

and [I am not completely sure](https://en.wikipedia.org/wiki/Goldbach%27s_conjecture) this will always run the `withMaven` step. :grin:

What you _can_ do is tell what steps actually _were_ run for a specific completed build, or have been run up until now for a specific running build—which is what you want to know here.

**Comment by bguerin** [comment]
Yes you are right

I got lost because the doc only talk about Job, not Build/Run

But I do know the differences between Job and Build/Run and between a FreestyleJob and a PipelineJob

Anyway, thanks again for your time and pointers !

**Comment by bguerin** [comment]
@jglick could you please have a look on my flow graph inspection code, to ensure I am doing it the right way ? Thanks !
(I keep in mind your remark about not using `WorkflowRun` directly, will change this later)

File: jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/listeners/AbstractWorkflowRunListener.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        return ofNullable(run.asFlowExecutionOwner())
+                .map(owner -> {
+                    try {
+                        return owner.get();
+                    } catch (Exception ex) {
+                        listener.getLogger().println("[withMaven] downstreamPipelineTriggerRunListener - Failure to introspect build steps: " + ex.toString());
+                        return null;
+                    }
+                })
+                .map(execution -> new FlowGraphWalker(execution))
```
Comment (jglick) [inline jenkins-plugin/src/main/java/org/jenkinsci/plugins/pipeline/maven/listeners/AbstractWorkflowRunListener.java:26]: Works, but `DepthFirstScanner` should be preferred.

**Comment by bguerin** [comment]
> if you already have a `WorkflowRun` via dep on `workflow-job`, which is a bad practice

fixed: https://github.com/jenkinsci/pipeline-maven-plugin/pull/517

**Pull request merged**